shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java

```diff
@@ -0,0 +1,131 @@
+     * @param storageDataSourceStates storage node data source status
+     * @param force whether to force start
+     */
+    public void initStates(final String databaseName, final Map<String, DataSource> dataSources, final Map<String, DataSourceState> storageDataSourceStates, final boolean force) {
+        this.force = force;
+        dataSources.forEach((key, value) -> {
+            initState(databaseName, storageDataSourceStates, key, value);
+        });
+        initialized = true;
+    }
+    
+    private void initState(final String databaseName, final Map<String, DataSourceState> storageDataSourceStates, final String actualDataSourceName, final DataSource dataSource) {
+        DataSourceState storageState = storageDataSourceStates.get(getCacheKey(databaseName, actualDataSourceName));
+        if (DataSourceState.DISABLED == storageState) {
+            dataSourceStates.put(getCacheKey(databaseName, actualDataSourceName), storageState);
+        } else {
+            checkState(databaseName, actualDataSourceName, dataSource);
+        }
+    }
+    
+    private void checkState(final String databaseName, final String actualDataSourceName, final DataSource dataSource) {
+        try (Connection ignored = dataSource.getConnection()) {
+            dataSourceStates.put(getCacheKey(databaseName, actualDataSourceName), DataSourceState.ENABLED);
+        } catch (final SQLException ex) {
+            if (this.force) {
+                log.error("Data source state unavailable, ignored with the -f parameter.", ex);
+            } else {
+                throw new DataSourceStateException("DataSourceState", 1, "Data source state unavailable.", ex);
+            }
+        }
+    }
+    
+    /**
+     * Get enabled data sources.
+     *
+     * @param databaseName database name
+     * @param databaseConfig database config
+     * @return enabled data sources
+     */
+    public Collection<DataSource> getEnabledDataSources(final String databaseName, final DatabaseConfiguration databaseConfig) {
+        return databaseConfig.getDataSources().isEmpty() ? Collections.emptyList() : getEnabledDataSourceMap(databaseName, databaseConfig.getDataSources()).values();
+    }
+    
+    /**
+     * Get enabled data source map.
+     *
+     * @param databaseName database name
+     * @param dataSources data sources
+     * @return enabled data source map
+     */
+    public Map<String, DataSource> getEnabledDataSourceMap(final String databaseName, final Map<String, DataSource> dataSources) {
+        if (dataSources.isEmpty() || !initialized) {
+            return dataSources;
+        }
+        Map<String, DataSource> result = new LinkedHashMap<>(dataSources.size(), 1);
+        dataSources.forEach((key, value) -> {
+            DataSourceState dataSourceState = dataSourceStates.get(getCacheKey(databaseName, key));
+            if (force ? DataSourceState.ENABLED == dataSourceState : DataSourceState.DISABLED != dataSourceState) {
```

Comment:
Please extract this logic to a param out of foreach loop.

Reply:
Ok